### PR TITLE
developed getExposure and setExposure; developed makeGif

### PR DESCRIPTION
**Pull request by AlexKotov**
some new functions that allows to make gif images from multiple images and save it in 480p;
also there is two methods that helps to change "brightness" of canvas and images;
also developed flash effect when taking photo like in native camera app

### CODE REVIEW

File: ios/RCT/RCTCameraManager.m [hunk 1]
```diff
@@ -10,16 +10,20 @@
...
 
 @property (strong, nonatomic) RCTSensorOrientationChecker * sensorOrientationChecker;
 @property (assign, nonatomic) NSInteger* flashMode;
+@property (nonatomic, strong) UIView *overlay;
 
 @end
 
 @implementation RCTCameraManager
 
+@synthesize overlay;
```
[the remark this answers is not in the document]
Comment (AlexKotov) [reply ios/RCT/RCTCameraManager.m]: i'll remove this and will push to master

**Comment by AlexKotov** [comment]
I'll migrate to RNCamera completely then add this features to new lib and
make pull request

2018-04-17 15:54 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

**Comment by sibelius** [comment]
I think you can add to both RCTCamera and RNCamera

could u add this to the docs as well?

**Comment by AlexKotov** [comment]
thanx.

2018-04-17 18:08 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

**Comment by sibelius** [comment]
@AlexKotov can u add those methods to RNCamera as well?

**Comment by AlexKotov** [comment]
I'm doing this now

2018-04-18 16:39 GMT+03:00 Sibelius Seraphini <notifications@github.com>:

**Comment by AlexKotov** [comment]
I see that pictures taken by RNCamera "falling" into some separate part of
memory, "file:///var/mobile/Containers/Data/Application/", not in Camera
Roll or Gallery. How can I configure this?

2018-04-18 16:52 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

**Comment by AlexKotov** [comment]
Or how can I get access to these pictures from React Native or from your lib

2018-04-18 17:32 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

**Comment by AlexKotov** [comment]
See this method from official React Native 0.55: saveToCameraRoll()
I can use it for my purposes

2018-04-18 17:35 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

**Comment by AlexKotov** [comment]
Used method. Now picture is in Camera Roll but it's rotated to the left
(90) :(

2018-04-18 17:39 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

**Comment by jgfidelis** [comment]
Have you tried using options like forceUpOrientation?

**Comment by AlexKotov** [comment]
forceUpOrientation helped. Thank u.

2018-04-18 19:55 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

**Comment by AlexKotov** [comment]
Think I need help with bringing improvements to RNCamera

I added to RNCameraManager.m lines:

RCT_REMAP_METHOD(getExposureCompensationRange,

                 reactTag:(nonnull NSNumber *)reactTag

                 callback:(RCTResponseSenderBlock)callback) {

  [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager,
NSDictionary<NSNumber *, RNCamera *> *viewRegistry) {

    RNCamera *view = viewRegistry[reactTag];

    if (![view isKindOfClass:[RNCamera class]]) {

      RCTLogError(@"Invalid view returned from registry, expecting
RNCamera, got: %@", view);

    } else {

      callback(@[[NSNull null], [view getExposureCompensationRange]]);

    }

  }];

}

I added to RNCamera.m lines:

- (NSArray *)getExposureCompensationRange

{

  AVCaptureDevice *backCamera = [[self videoCaptureDeviceInput] device];



  float min = backCamera.minExposureTargetBias;

  float max = backCamera.maxExposureTargetBias;



  NSNumber *minNumber = [NSNumber numberWithFloat:min];

  NSNumber *maxNumber = [NSNumber numberWithFloat:max];

  id objects[] = {minNumber, maxNumber};

  NSUInteger count = sizeof(objects) / sizeof(id);



  return [NSArray arrayWithObjects:objects count:count];

}

I added to RNCamera.h line:

- (NSArray *)getExposureCompensationRange;

I changed JS code from:
NativeModules.CameraManager
to:
NativeModules.RNCameraManager

When I'm using getExposureCompensationRange from JS I have an error:
'NativeModules.RNCameraManager.getExposureCompensationRange was called with
1 arguments but expects 2 arguments'

2018-04-19 12:14 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

**Comment by jgfidelis** [comment]
Don't change the JS code, NativeModules.CameraManager, this is already like this for a reason.

How are you calling `getExposureCompensationRange`? Check the methods that are already being called to see how it should be. You should pass the cameraHandle to be the `reactTag` parameter

**Comment by AlexKotov** [comment]
If I don't change JS code then wrong methods would be callen (methods from
old lib - RCT, not from new one - RN)

2018-04-20 18:55 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

**Comment by jgfidelis** [comment]
what js file are you changing?

**Comment by AlexKotov** [comment]
what do u mean?

2018-04-20 19:44 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

**Comment by AlexKotov** [comment]
ok. Now i'm using NativeModules.CameraManager. I had modified RNCamera.js
and index.d.ts but getExposureCompensationRange from RCTCameraManager.m was
called.
I have problem with understanding of bridging in your lib.

2018-04-20 20:51 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

**Comment by n1ru4l** [comment]
Going to close this since RCTCamera is deprecated and be removed in the next major release. Please feel free to implement these feature for RNCamera. 

Please also try to put only a single feature in to one pull request to reduce review complexity.

**Pull request closed without merging**